src/manager.rs

```diff
@@ -407,6 +354,128 @@ impl ManagerService for Arc<Manager> {
     }
 }
 
+fn compute_quorum_results(
+    replica_id: &str,
+    rank: i64,
+    quorum: &Quorum,
+) -> Result<ManagerQuorumResponse, Status> {
+    let mut participants = quorum.participants.clone();
+    participants.sort_by(|a, b| a.replica_id.cmp(&b.replica_id));
+
+    // Compute the rank of the replica in the returned quorum.
+    let replica_rank = participants
+        .iter()
+        .enumerate()
+        .find_map(|(i, p)| {
+            if p.replica_id == replica_id {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .ok_or_else(|| {
+            Status::not_found(format!(
+                "replica {} not participating in returned quorum",
+                replica_id
+            ))
+        })?;
+
+    let step = participants[replica_rank].step;
+
+    // Compute the details for workers at max step.
+    let max_step = participants.iter().map(|p| p.step).max().unwrap();
+    let max_participants: Vec<&QuorumMember> =
+        participants.iter().filter(|p| p.step == max_step).collect();
+    let max_rank = max_participants.iter().enumerate().find_map(|(i, p)| {
+        if p.replica_id == replica_id {
+            Some(i as i64)
+        } else {
+            None
+        }
+    });
+
+    // The primary TCPStore to use for this rank.
+    let primary_rank = rank as usize % max_participants.len();
+    let primary = max_participants[primary_rank];
+
+    // Compute recovery assignments
+
+    // Nodes are recovering if:
+    // 1. not at the max step
+    // 2. max_step == 0 and not the primary replica
+    let all_recovering_ranks: Vec<usize> = participants
+        .iter()
+        .enumerate()
+        .filter_map(|(i, p)| {
+            if p.step != max_step || max_step == 0 && primary.replica_id != p.replica_id {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect();
+    let all_recovering_ranks_set = all_recovering_ranks.iter().collect::<HashSet<_>>();
+    let serving_ranks: Vec<usize> = participants
```

Comment:
nit: I was initially confused by the name serving_ranks, maybe "stable_ranks" or "up_to_date_ranks" would also make sense?

Reply:
done